### PR TITLE
[new module] Added Azure SQL Servers support - cloud/azure/azure_rm_sql_servers

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Adding support for Azure SQL Servers

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
azure_rm_sql_servers

##### ANSIBLE VERSION
```
ansible 2.5.0 (devel 125c4b135d) last updated 2017/10/17 07:30:02 (GMT +000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /ansible/lib/ansible
  executable location = /ansible/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: short_description should not have a `.`, are you able to expand on this like, `Manage a Azure SQL Server`?

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: SQL

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: It would be nice if this wasn't required and could be derived from the resource group specified.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: `required: False` is the default and is usually omitted.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: I would argue that `SystemAssigned` should be the default value if this is not set or can a SQL Server not have an identity? Is there a reason why this is a nested option under the `type` key? If not can identity just by the type value or do you see there being more options available to identity?

When quoting a value, you can enclose it with `C()`, e.g. `'SystemAssigned'` should be `C(SystemAssigned)`. You can also use `I()` when referring to option option names.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: not needed

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Change this option to `admin_username` to align with the other Azure modules.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Change this to `administrator_password` to align with other Azure modules.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Not needed

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Not needed

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Not needed

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Can you expand on this more, e.g. how the default is derived if omitted, where to find a list of valid versions and so on.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Is this the same with `administrator_username`, can it not be changed? Should there be another parameter like `update_password` which only makes a change on creation or all the time, see the `user` module for more details http://docs.ansible.com/ansible/latest/user_module.html.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Can you keep this left aligned (remove the 4 spaces), also try and avoid using `{{}}` blocks and put valid values in there, it makes it easier for newer people to see what the values can actually be.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Would be nice if you can actually show an example of a valid version value, nothing worse than seeing that it can be specified but not knowing the form it takes.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: We should start looking at stopping this behaviour and returning relevant values in the root level of the return. When testing this module I see the following is returned

```
"state": {
    "administrator_login": "username",
    "fully_qualified_domain_name": "jborean-sql-server.database.windows.net",
    "id": "/subscriptions/3f7e29ba-24e0-42f6-8d9c-5149a14bda37/resourceGroups/jborean-testgroup/providers/Microsoft.Sql/servers/jborean-sql-server",
    "kind": "v12.0",
    "location": "australiaeast",
    "name": "jborean-sql-server",
    "state": "Ready",
    "type": "Microsoft.Sql/servers",
    "version": "12.0"
}
```

So let's do the following and create a return value for:
* fqdn
* id
* version

I don't think the other values are really necessary and are more suited towards a facts module if we want to gather that information. I decided to omit the `location`, `admin login`, `kind, `name`, `state` and `type` as they would be ones you set in the module parameters and already know. Version is a tiny bit different as you can omit it and I could argue `type` could potentially be returned if you can specify different types in the future.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Add another entry `no_log=True` to stop Ansible from logging this value, this should be set on any secret parameters (like passwords) to keep it a secret.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: Have a look at [azure_rm_virtualmachine](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/azure/azure_rm_virtualmachine.py) to see how to do this, you usually set the parameter as None in the __init__ and these are populated in `AzureRMModuleBase` for you.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: This shouldn't be run in check mode.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: You need some code to check if the server actually needs updating, right now the module is reporting a change when running it twice which is not idempotent.

File: lib/ansible/modules/cloud/azure/azure_rm_sql_server.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sql_server.py]: We should look at changing the return values, see my note in the documentation portion.

File: test/integration/targets/azure_rm_sql_server/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/azure_rm_sql_server/tasks/main.yml]: I know you were saying you were having issues trying to get this to run but we should also look at expanding these tests a lot more, usually we look at performing 3 tasks per action, e.g. 3 tasks for creating a server, 3 tasks for modifying a server and 3 tasks for removing a server. The 3 tasks per action should follow;

* Run the task in check mode and check that the server wasn't actually created
* Run the task normally and check that the server was created as expected
* Run the task again (idempotent) and check that a change wasn't reported

**Review by jborean93 — CHANGES_REQUESTED** [review]
Have added some comments, some things we would need to do before merging is

* Fix up the idempotency, right now it reports a change everytime it runs
* Expand the tests to validate the module
* Make the return values more Ansible like

Overall the module works for me and creates a SQL Server which is great. One question I do have more on the Azure side, is it planned to add support for more SQL types, e.g. MySQL and so or is that not supported in Azure or going to be in separate modules?

**Comment by zikalino** [comment]
ready_for_review

**Comment by zikalino** [comment]
fixed several major issues:
- renaming, etc.
- idempotence support
- added update / idempotence tests
I may have missed some smaller issues.
Please recheck if still any important changes needed

Additional note - I still plan to add following adjustments in this PR:
- flatten state in response (like in Contatiner Registry module)
- add default values
- flatten parameters

**Comment by zikalino** [comment]
ready_for_review

**Comment by zikalino** [comment]
@jborean93 please check the module again, i think it's ready for merging.

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: I'm pretty sure you don't need the tags as you are extending the `azure_tags` documentation below.

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:171]: What's the point of this whole block, `AzureRMModuleBase` should be setting `self.x` based on the module_arg_spec

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: No need for a `try/except` here as azure_rm_common handles a missing source group

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: This doesn't seem to be working, I got `The error was: msrest.exceptions.ValidationError: Parameter 'Server.location' can not be None` in an error which means this musn't work. Try doing;

```
if self.location is None:
    self.location = resource_group.location
```

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: `state` doesn't need to be there.

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: Instead of updating the results object here with the response and then removing the unneeded ones below, just set the results you want manually.

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: Need to not run when in check mode

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (jborean93) [inline lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py]: This shouldn't be returned, it should be the same values that you set in add or update.

File: test/integration/targets/azure_rm_sqlserver/aliases
Comment (jborean93) [inline test/integration/targets/azure_rm_sqlserver/aliases:2]: You will need to add `posix/ci/cloud/group2/azure` to this list so the tests actually run in CI

**Review by jborean93 — CHANGES_REQUESTED** [review]
I couldn't really test this as I came across the error

`fatal: [127.0.0.1]: FAILED! => {"changed": false, "msg": "Error creating the SQL Server instance: Azure Error: InvalidRequestContent\nMessage: The request content was invalid and could not be deserialized: 'Required property 'type' not found in JSON. Path 'identity', line 1, position 44.'."}`

A few things to note;

* The tests need to be activated by adding the alias I mentioned in the comment (that should have picked up the error I recieved)
* The tests need to be expanded to cover a check mode, normal and idempotent tests of the actions `create`, `update`, and `delete` scenarios
* I'm not sure if the tags are actually supported, I was unable to test it but couldn't find anything that really seem to check and set them
* The location is still required, it fails if you don't set it

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:171]: well, i have tried and it doesn't work as expected, so i have to leave it for now...

Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:171]: Ok, will try and debug and find out what is happening, it should just work.

Comment (jborean93) [reply lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:171]: Ok I see my mistake, most modules just have

```
for key in list(self.module_arg_spec.keys()) + ['tags']:
    setattr(self, key, kwargs[key])
```

Is there a reason why you have all these if statements as the above should be enough for you.

**Comment by zikalino** [comment]
@jborean93 I have made all necessary updates, seems like test can't run as Microsoft.Sql namespace has to be added.

Error creating the SQL Server instance: Azure Error: MissingSubscriptionRegistration\nMessage: The subscription is not registered to use namespace 'Microsoft.Sql'. See https://aka.ms/rps-not-found for how to register subscriptions.\nException Details:\n\tError Code: MissingSubscriptionRegistration\n\tMessage: The subscription is not registered to use namespace 'Microsoft.Sql'. See https://aka.ms/rps-not-found for how to register subscriptions.\n\tTarget: Microsoft.Sql

**Comment by jborean93** [comment]
Looks like the tests are failing because we aren't authorised in CI, will see if @mattclay can help.

Edit: just saw your post :)

**Comment by jborean93** [comment]
@zikalino once we get the tests running we need to look at expanding the tests, there is at least 1 issue I can find that the tests should have picked up if we had some more tests. Let me know if you are confused as to what I am asking for, I'm happy to discuss it further over IRC as it would be good to get this merged today so you can update your other PR's with the changes you made here.

File: lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_sqlserver.py:171]: yes, these attributes are sticked into objects padded to Azure SDK, so treated in different way

**Comment by zikalino** [comment]
@jborean do you mean taking location from resource group? this could be difficult. some resources can't be created in particular location (I believe i couldn't create SQL server in the same location as the one where default rg is created).

**Comment by jborean93** [comment]
@zikalino, for the location part I see the benefit in being able to set the location independently to override the resource group but most modules seem to have it as an optional value and if not set explicitly it is derived from the resource group location.

What I am talking about is expanding the tests from just the create, update, delete tasks to test each scenario under check mode, normal and an idempotent task. In the end it would look similar to

```
- name: create sql server (check)
  azure_rm_sqlserver:
    ...
  register: create_server_check

- name: check if server exists (check)
  ... ? use a facts module if possible
  register: create_server_result_check

- name: assert create sql server results (check)
  assert:
    that:
    - create_server_check is changed
    - create_server_check.exists == False

- name: create sql server
  azure_rm_sqlserver:
    ...
  register: create_server

- name: check if server exists
  ... ? use a facts module if possible
  register: create_server_result

- name: assert create sql server results
  assert:
    that:
    - create_server is changed
    - create_server.exists == True

- name: create sql server (idempotent)
  azure_rm_sqlserver:
    ...
  register: create_server_again

- name: assert create sql server results (idempotent)
  assert:
    that:
    - not create_server_again is changed
```

This would be repeated again for updating a SQL server and then finally again for removing a SQL server.

**Comment by jborean93** [comment]
@zikalino, the Microsoft.SQL resource has been registered and the tests are now running.

**Comment by zikalino** [comment]
@jborean93 seems now tests are passing. should i still add more tests to this pr? or is it ok to merge, and i could expand tests when adding facts module?

**Comment by jborean93** [comment]
I think we should still continue to add the extra tests just without that second step of checking the results with a fact module for now (should still come back and expand once a facts module is in). We can still do some assertions on the results of just the task itself.

Do you know how to run tests locally?

**Comment by zikalino** [comment]
@jborean93 I have more tests for check_mode and also delete unexisting sql server test (which i moved to front, as delete after delete seems a bit unstable -- second delete may get invalid information that resource still exists)

**Comment by jborean93** [comment]
Looks better, we still need a test for each action again to ensure that no changes are recorded (idempotent).

**Comment by zikalino** [comment]
@jborean93 how about now? added 2 more tests

File: test/integration/targets/azure_rm_sqlserver/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/azure_rm_sqlserver/tasks/main.yml]: Delete this one, you only need to run the check mode test for a scenario once

File: test/integration/targets/azure_rm_sqlserver/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/azure_rm_sqlserver/tasks/main.yml]: Take away this test

File: test/integration/targets/azure_rm_sqlserver/tasks/main.yml
Comment (jborean93) [inline test/integration/targets/azure_rm_sqlserver/tasks/main.yml]: Don't need this test as you should be starting with a random resource group so no SQL server would exist anyway

**Review by jborean93 — COMMENTED** [review]
Looking good, left some comments on some tests, only thing remaining is to delete the instance again right at the end for idempotency checks

**Review by jborean93 — APPROVED** [review]

**Comment by zikalino** [comment]
shipit

**Comment by jborean93** [comment]
Awesome @zikalino, are you able to raise a PR to add this module to the changelog for 2.5 https://github.com/ansible/ansible/blob/devel/CHANGELOG.md. We try to keep things in alphabetical order in each section.

**Pull request merged**
